Adding bash and zsh tab-completion

coderforlife opened the pull request:
This code adds thorough bash and zsh tab-completion for the gkeep client, closing #113.

It looks as though you are trying to use the `argcomplete` package to do this, however it has issues:

* can only complete static commands, it has no idea how to get a list of classes or students in the appropriate places
* it does not work on modern macOS - it assumes bash completions are at /etc/bash_completion.d however they are at `/opt/homebrew/Cellar/bash-completion@2/2.11/share/bash-completion/completions` or similar (obtainable with `pkg-config --variable=completionsdir bash-completion`)

This pull request provides a much better experience as it can fill in all arguments appropriately by using data from the server and similar.

The integration it uses is modeled after `pip`'s completion system.

This only has some small modifications to the `gkeep` client: removing argcomplete usage, adding a short-circuit if it detects that auto-completion is requested, and a completion sub-command to generate the code for integrating the autocompletion (or even installing it in the most common setups). All other code is in a self-contained file.

coderforlife committed:
Remove argcomplete

coderforlife committed:
Adding core completion code

At this point completion works but must be installed completely manually.

coderforlife committed:
Add code to install code completion for bash and zsh

bjcoleman commented on git-keeper-client/gkeepclient/completion.py:
```diff
@@ -0,0 +1,688 @@
+    # COMP_WORDS: an array of all the words in the command line
+    # and others...
+
+    # # Get command line using COMP_LINE and COMP_POINT
```
Please remove old code.

bjcoleman commented:
Thank you for taking the time to develop this.  It will be a great addition!

Can you add section in `docs/faculty-users.md` with setup and (if needed) usage instructions?

coderforlife committed:
Removing old code.

coderforlife commented:
Should the docs come just after the *Help* section?

bjcoleman commented:
Yes, that looks like a good spot.  That's at the end of the setup portion, which makes sense for something that is optional.

coderforlife committed:
Adding documentation for tab completion

coderforlife commented:
Addressed both issues.

coderforlife committed:
Merge branch 'develop' into develop

coderforlife commented:
Can this be revisited?

sommern commented:
This will be a very welcome addition!

Some comments and issues:

The `--install` switch for bash installed here on my Macbook:

```
/Users/sommern1/.local/share/bash-completion/completions/gkeep
```

I then needed to install and set up `bash-completion@2` using homebrew for it to work. The docs recommend having bash-completion installed but it is not required. Is my assumption correct that it is required for `--install` to work, but without it one needs to edit their bash configuration manually? If so, this detail should be in the docs.

For zsh, the configuration has `\v` characters which get turned into actual vertical tabs in the output. Replacing both `\v` instances with `\\v` on line 526 of completion.py fixed it.

I was not able to get this working with oh-my-zsh and `--install`. I created a fresh user on my Linux machine, installed oh-my-zsh as that user, and then ran `gkeep completion --install zsh` which installed in `~/.oh-my-zsh/completions/__gkeep`, but completion did not work. When I put the same contents in `~/.zshrc` it worked just fine.

I think `--install` should ask for confirmation so that the user can review the destination path before it is actually written. There is a `confirmation()` function in `text_ui.py` you could use.

If the server cannot be reached or if there is a syntax error in the client config file, attempting completion that requires fetching info from the server causes a confusing stack trace rather than a helpful error message.

For future maintenance it would be nice to have some information in the `completion.py` docstring about how to add completion for a new subcommand.

coderforlife committed:
Fix literal vertical tab.

coderforlife committed:
Improve bash completions install script

coderforlife committed:
Improving zsh completion slightly.

coderforlife committed:
Adding notes on adding new completions for shells.

coderforlife commented:
Requirement of bash-completions: in the code I actually wrote "The bash-completions package is
 required to be installed for this to be useful." There is also a TODO about making it so it supports things when you don't have bash-completions installed. I have improved this now and if it detects bash-completions is not installed, it warns the user and asks to install to the bashrc file instead.

zsh and `\v`: fixed

oh-my-zsh: don't use zsh, so not sure, I will look into it more soon

Asking for confirmation: added.

Stack traces: using bash on linux or mac, I cannot reproduce; is this your experience? I will try more with zsh soon

Adding new subcommands: info added.
